### PR TITLE
Implement multi-scale detection and deblending algorithms

### PR DESCRIPTION
**Pull request by fred3m**
This PR contains a number of improvements:
- A new `MonotonicMaskConstraint` has been created to allow more flexible monotonicity
- The a trous algorithm to apply the starlet transforms has been implemented to improve performance
- Multi-scale detection using starlets has been implemented

### CODE REVIEW

File: scarlet/wavelet.py [hunk 1]
```diff
@@ -16,304 +11,369 @@ class Starlet(object):
...
-        self._starlet_shape = [self._lvl, *self._image_shape[-2:]]
-        if self.seed is None:
-            self.seed = mk_starlet(self._starlet_shape)
-        self._norm = np.sqrt(np.sum(self.seed ** 2, axis=(-2, -1)))
+        self._generation = generation
+        self._convolve2D = convolve2D
+        self._norm = None
+
+    @staticmethod
+    def fromImage(image, scales=None, generation=2, convolve2D=None):
```
Comment (pmelchior) [inline scarlet/wavelet.py:36]: I was always a fan of CamelCase, but for consistency let's call this thing `from_image`

File: scarlet/wavelet.py [hunk 2]
```diff
@@ -16,304 +11,369 @@ class Starlet(object):
...
+            The resulting `Starlet` that contains the image, starlet coefficients,
+            as well as the parameters used to generate the coefficients.
+        """
+        if scales is None:
+            scales = get_scales(image.shape)
+        coefficients = starlet_transform(image, scales, generation, convolve2D)
+        return Starlet(image, coefficients, generation, convolve2D)
+
+    @staticmethod
+    def fromCoefficients(coefficients, generation=2, convolve2D=None):
```
Comment (pmelchior) [inline scarlet/wavelet.py:66]: -> `from_coefficients`

**Review by pmelchior — COMMENTED** [review]
Really minor things: no camelCase. I made some changes to the wavelet tutorial as well.

File: scarlet/wavelet.py (hunk 1 above)
Comment (fred3m) [reply scarlet/wavelet.py:36]: Yes, I've just been writing a lot of LSST code lately and got into the habit of writing camel case again (which I prefer too, but is not standard python).

**Review by pmelchior — APPROVED** [review]

**Pull request merged**